feat(linter): implement no-restricted-types rule

camc314 opened the pull request:
Implements the typescript-eslint no-restricted-types rule which allows
users to ban specific types from being used. Supports all TypeScript
keyword types (including `any`), custom types, namespaced types,
generic types, empty tuples, and empty object types.

Includes fixWith for auto-fixes and suggest for manual suggestions.

connorshea commented on crates/oxc_linter/src/rules/typescript/no_restricted_types.rs:
```diff
@@ -0,0 +1,557 @@
+    NoRestrictedTypes,
+    typescript,
+    restriction,
+    fix_suggestion,
```
Should fix the test failure:
```suggestion
    fix_suggestion,
    config = NoRestrictedTypesConfig,
```

camc314 replied:
lol thanks. i thought i did this, clearly not

camc314 commented on crates/oxc_linter/src/rules/typescript/no_restricted_types.rs:
```diff
@@ -0,0 +1,557 @@
+    AstNode,
+    context::{ContextHost, LintContext},
+    rule::Rule,
+};
```
?? it looks like that already?

connorshea commented on crates/oxc_linter/src/rules/typescript/no_restricted_types.rs:
```diff
@@ -0,0 +1,558 @@
+        Self(Box::new(config))
+    }
+
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
```
Noob question so possibly irrelevant, but: Is this matcher sufficiently performant? e.g. should we do anything to pre-calculate which ast kinds to check based on the banned types, rather than doing that check every time there's a node of any of these types?

connorshea replied:
Also, would this automatically skip these checks for an empty array of banned types? And can this be implemented as `run_once`?

camc314 commented:
> Is it a concern that the user can theoretically cause a fix loop, e.g. ban `BadType` and set `fixWith: BadType` (or a more complex one where they have `BadType` and `WorseType` and set them to fix into one another? They'd have to be dumb, but I guess it's possible. It's probably fine if they decide to footgun themselves, but figured I'd note it.

eh that's just a config issue I think, users just shouldn't configure it like that.

connorshea commented:
Unfortunately it looks like the configuration docs still don't get serialized quite right here

camc314 commented on crates/oxc_linter/src/rules/typescript/no_restricted_types.rs:
```diff
@@ -0,0 +1,560 @@
+    fn from_configuration(value: serde_json::Value) -> Self {
+        let config = value
+            .get(0)
+            .and_then(|v| serde_json::from_value::<NoRestrictedTypesConfig>(v.clone()).ok())
```
fixed

camc314 committed:
feat(linter): implement no-restricted-types rule (#16276)

Implements the typescript-eslint no-restricted-types rule which allows
users to ban specific types from being used. Supports all TypeScript
keyword types (including `any`), custom types, namespaced types,
generic types, empty tuples, and empty object types.

Includes fixWith for auto-fixes and suggest for manual suggestions.

Perdolique commented:
Ahh, okay. I waited for a few hours after the release announcement, but it still wasn't there, so I thought it might be a bug. It works now 👍
